### PR TITLE
Add window/didChangeFocus notification calls 

### PR DESCRIPTION
**Pull request by mkondratek**
Related PR: https://github.com/sourcegraph/cody/pull/5997

## Test plan
I reviewed the `cody-agent-trace.json`. The requests appear correctly in proper time. Also, I checked the agent in a debugger session - the requests come and have the proper param.

### CODE REVIEW

File: CONTRIBUTING.md
Comment (mkondratek) [inline CONTRIBUTING.md:1]: unrel fixes - GitHub does not handle \n correctly

File: src/main/kotlin/com/sourcegraph/cody/agent/CodyAgentServer.kt [hunk 1]
```diff
@@ -79,8 +87,15 @@ interface _SubsetGeneratedCodyAgentServer {
...
 
+  @JsonNotification("initialized") fun initialized(params: Null?)
+
+  @JsonNotification("exit") fun exit(params: Null?)
+
   @JsonNotification("extensionConfiguration/didChange")
   fun extensionConfiguration_didChange(params: ExtensionConfiguration)
+
+  @JsonNotification("window/didChangeFocus")
+  fun window_didChangeFocus(params: Window_DidChangeFocusParams)
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/agent/CodyAgentServer.kt:98]: this is the new notification - the others have been reorganized only

File: src/main/kotlin/com/sourcegraph/cody/listeners/CodyFocusChangeListener.kt [hunk 2]
```diff
@@ -15,8 +16,15 @@ class CodyFocusChangeListener(val project: Project) : FocusChangeListener {
       EditorChangesBus.documentChanged(project, textDocument)
       CodyAgentService.withAgent(project) { agent: CodyAgent ->
         agent.server.textDocumentDidFocus(textDocument)
+        agent.server.window_didChangeFocus(Window_DidChangeFocusParams(true))
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/listeners/CodyFocusChangeListener.kt]: Here, we perform two calls:
- `textDocument/didFocus`
- `window/didChangeFocus`

I wonder... do we need the two? maybe we can optimize it 🤔

Comment (mkondratek) [reply src/main/kotlin/com/sourcegraph/cody/listeners/CodyFocusChangeListener.kt]: I think `textDocument/didFocus` notification implies `window/didChangeFocus` (true).

File: src/main/kotlin/com/sourcegraph/cody/listeners/CodyFocusChangeListener.kt [hunk 3]
```diff
@@ -15,8 +16,15 @@ class CodyFocusChangeListener(val project: Project) : FocusChangeListener {
...
       }
       IgnoreOracle.getInstance(project).focusedFileDidChange(textDocument.uri)
     }
   }
+
+  override fun focusLost(editor: Editor) {
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/listeners/CodyFocusChangeListener.kt]: Switching b/w files in the project calls `focusLost` and `focusGained` immediately afterwards.

Is it ok?

File: src/main/kotlin/com/sourcegraph/cody/listeners/CodyFocusChangeListener.kt (hunk 2 above)
Comment (pkukielka) [reply src/main/kotlin/com/sourcegraph/cody/listeners/CodyFocusChangeListener.kt]: That is true for focus gain, but not for the focus lost.
But if we need notification about focus lost or not - I cannot say :)

File: src/main/kotlin/com/sourcegraph/cody/listeners/CodyFocusChangeListener.kt (hunk 3 above)
Comment (pkukielka) [reply src/main/kotlin/com/sourcegraph/cody/listeners/CodyFocusChangeListener.kt]: I think that is incorrect implementation:
1. That focus `flickering` is incorrect on its own
2. What if user will close the only one opened document?

What you need is new listener:

```
    val frame = WindowManager.getInstance().getFrame(project)
    val listener =
        object : WindowAdapter() {
          override fun windowActivated(e: WindowEvent?) {
            super.windowActivated(e)
            ApplicationManager.getApplication().executeOnPooledThread { getAuthenticationState() }
          }
        }
    frame?.addWindowListener(listener)
    Disposer.register(this) { frame?.removeWindowListener(listener) }
```

This code is copied, please take a look at `CodyAuthenticationManager::addAuthChangeListener`. You would also need to override `windowDeactivated`

Comment (mkondratek) [reply src/main/kotlin/com/sourcegraph/cody/listeners/CodyFocusChangeListener.kt]: done

File: src/main/kotlin/com/sourcegraph/cody/config/CodyAuthenticationManager.kt [hunk 4]
```diff
@@ -114,6 +116,16 @@ class CodyAuthenticationManager :
           override fun windowActivated(e: WindowEvent?) {
             super.windowActivated(e)
             ApplicationManager.getApplication().executeOnPooledThread { getAuthenticationState() }
+            CodyAgentService.withAgent(project) { agent: CodyAgent ->
```
Comment (pkukielka) [inline src/main/kotlin/com/sourcegraph/cody/config/CodyAuthenticationManager.kt]: Can we extract this listener somewhere an then subscribe to it (e.g. have option to add a callback) if we want in places we need? Or even we can hardcode call to `CodyAccountManager` inside it if you want to make it simpler, structure will just have more sense other way around.
Doing general state notifications inside `CodyAccountManager` looks weird...

Comment (mkondratek) [reply src/main/kotlin/com/sourcegraph/cody/config/CodyAuthenticationManager.kt]: I extracted the class

File: src/main/kotlin/com/sourcegraph/cody/config/CodyWindowManager.kt [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.project.Project
+import com.sourcegraph.cody.agent.CodyAgent
+import com.sourcegraph.cody.agent.CodyAgentService
+import com.sourcegraph.cody.agent.protocol_generated.Window_DidChangeFocusParams
+import java.awt.event.WindowAdapter
+import java.awt.event.WindowEvent
+
+class CodyWindowAdapter(private val project: Project) : WindowAdapter() {
```
Comment (pkukielka) [inline src/main/kotlin/com/sourcegraph/cody/config/CodyWindowManager.kt]: Can you also rename file to match class name?

Comment (mkondratek) [reply src/main/kotlin/com/sourcegraph/cody/config/CodyWindowManager.kt]: done

**Review by pkukielka — APPROVED** [review]
LGTM

File: src/main/kotlin/com/sourcegraph/cody/config/CodyAuthenticationManager.kt [hunk 6]
```diff
@@ -109,21 +107,15 @@ class CodyAuthenticationManager :
 
   fun addAuthChangeListener(project: Project) {
```
Comment (pkukielka) [inline src/main/kotlin/com/sourcegraph/cody/config/CodyAuthenticationManager.kt]: This method should be now part of `CodyWindowAdapter`, it does not make sense in `CodyAuthenticationManager`

**Pull request merged**